### PR TITLE
[k8s] ensure agent shuts down on critical watch failures.

### PR DESCRIPTION
**Pull request by darobs**
Sometimes the initial List in a List/Watch API request will time out, and the exception is swallowed instead of being handled.
This is what happened in issues like #3764 .

The main program has a CancellationTokenSource, created to shut down the main process from SIGTERM events and other critical failures. This is now passed to the k8s watches, so the watch can cancel the CTS when a fatal error occurs.

The list of fatal errors:
- OnEvent: the event handler receives a fatal exception.
- OnClosed: An exception is thrown while attempting watch restart.
- OnError: Always triggers shutdown.

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironmentOperator.cs [hunk 1]
```diff
@@ -37,43 +38,67 @@ public void Stop()
...
         public void Dispose() => this.Stop();
 
-        void StartListPods() =>
+        void StartListPods(CancellationTokenSource shutdownCts) =>
             this.client.ListNamespacedPodWithHttpMessagesAsync(this.deviceNamespace, watch: true)
-                .ContinueWith(this.OnListPodsCompleted);
+                .ContinueWith(this.OnListPodsCompleted, shutdownCts);
 
-        async Task OnListPodsCompleted(Task<HttpOperationResponse<V1PodList>> task)
+        async Task OnListPodsCompleted(Task<HttpOperationResponse<V1PodList>> task, object shutdownCtsObject)
```
Comment (pmzara) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironmentOperator.cs:45]: Why the parameter shutdownCtsObject is an object and not CancellationTokenSource?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/EdgeDeploymentOperator.cs [hunk 2]
```diff
@@ -51,36 +51,23 @@ public void Stop()
...
-                        Events.EdgeDeploymentWatchFailed(ex);
-                        throw ex;
-                    }));
+                    onEvent: async (type, item) => await this.EdgeDeploymentOnEventHandlerAsync(type, item, shutdownCts),
+                    onClosed: () => this.RestartWatch(shutdownCts),
+                    onError: (ex) => this.HandleError(ex, shutdownCts)));
         }
 
-        internal async Task EdgeDeploymentOnEventHandlerAsync(WatchEventType type, EdgeDeploymentDefinition item)
+        internal async Task EdgeDeploymentOnEventHandlerAsync(WatchEventType type, EdgeDeploymentDefinition item, Option<CancellationTokenSource> shutdownCts)
```
Comment (pmzara) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/EdgeDeploymentOperator.cs:70]: Why use Option<CancellationTokenSource> and not only CancellationTokenSource? In some places, you are using CancellationTokenSource and in others, you are using Option<CancellationTokenSource>. Any specific reason to do this?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironmentOperator.cs (hunk 1 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironmentOperator.cs:45]: The [ContinueWith](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.continuewith?view=net-5.0#System_Threading_Tasks_Task_ContinueWith_System_Action_System_Threading_Tasks_Task_System_Object__System_Object_) on line 43 function is defined as `Action<Task,Object>` I'm probably being overly cautious. about the parameter type.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/EdgeDeploymentOperator.cs (hunk 2 above)
Comment (darobs) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/EdgeDeploymentOperator.cs:70]: I'm using the Option type to manage the potentially null object passed into `OnListEdgeDeploymentsCompleted`.  Logically, it shouldn't ever be null, but I think this is once again me being overly cautious.

**Review by pmzara — APPROVED** [review]

**Pull request merged**